packaging/release.py

```diff
@@ -0,0 +1,1420 @@
+        upstream_repo=git_state.remotes.upstream.repo,
+        upstream_branch=upstream_branch,
+        user=git_state.remotes.fork.user,
+        repo=git_state.remotes.fork.repo,
+        branch=branch,
+        title=title,
+        body=body,
+    )
+
+    return pr
+
+
+def create_github_release(release: GitHubRelease) -> None:
+    """Open a browser tab for creating the given GitHub release."""
+    # See: https://docs.github.com/en/repositories/releasing-projects-on-github/automation-for-release-forms-with-query-parameters
+
+    params = dict(
+        tag=release.tag,
+        target=release.target,
+        title=release.title,
+        body=release.body,
+        prerelease=1 if release.pre_release else 0,
+    )
+
+    query_string = urllib.parse.urlencode(params)
+    url = f"https://github.com/{release.user}/{release.repo}/releases/new?{query_string}"
+
+    display.show("Opening release creation page in new tab using default browser ...")
+    webbrowser.open_new_tab(url)
+
+
+def create_pull_request(pr: PullRequest) -> None:
+    """Open a browser tab for creating the given pull request."""
+    # See: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/using-query-parameters-to-create-a-pull-request  # noqa
+
+    params = dict(
+        quick_pull=1,
+        title=pr.title,
+        body=pr.body,
+    )
+
+    query_string = urllib.parse.urlencode(params)
+    url = f"https://github.com/{pr.upstream_user}/{pr.upstream_repo}/compare/{pr.upstream_branch}...{pr.user}:{pr.repo}:{pr.branch}?{query_string}"
+
+    display.show("Opening pull request in new tab using default browser ...")
+    webbrowser.open_new_tab(url)
+
+
+def create_pull_request_body(title: str) -> str:
+    """Return a simple pull request body created from the given title."""
+    body = f"""
+##### SUMMARY
+
+{title}
+
+##### ISSUE TYPE
+
+Feature Pull Request
+
+##### COMPONENT NAME
+
+ansible
+"""
+
+    return body.lstrip()
+
+
+def get_remote(name: str, push: bool) -> Remote:
+    """Return details about the specified remote."""
+    remote_url = git("remote", "get-url", *(["--push"] if push else []), name, capture_output=True).stdout.strip()
+    remote_match = re.search(r"[@/]github[.]com[:/](?P<user>[^/]+)/(?P<repo>[^.]+)(?:[.]git)?$", remote_url)
+
+    if not remote_match:
+        raise RuntimeError(f"Unable to identify the user and repo in the '{name}' remote: {remote_url}")
+
+    remote = Remote(
+        name=name,
+        user=remote_match.group("user"),
+        repo=remote_match.group("repo"),
+    )
+
+    return remote
+
+
+@functools.cache
+def get_remotes() -> Remotes:
+    """Return details about the remotes we need to use."""
+    # assume the devel branch has its upstream remote pointing to the user's fork
+    fork_remote_name = git("branch", "--list", "devel", "--format=%(upstream:remotename)", capture_output=True).stdout.strip()
```

Comment:
This assumption is broken for me:
```console
$ git branch --list devel '--format=%(upstream:remotename)'
origin

$ git branch --list packaging/pep517-in-tree-backend-mutation-tests '--format=%(upstream:remotename)'
fork
```

Reply:
@webknjaz Your `devel` branch remote points at ansible/ansible instead of your fork?

Reply:
Yep. This is because my userspace automation clones a lot of repos in a loop from their canonical locations, so they all have `origin` pointing upstream: https://github.com/webknjaz/ansible-gentoo-laptop/blob/e71ff8f/roles/gentoo-userspace/defaults/main.yml#L252-L573. It then adds a `fork` remote to select repos. It's worked like this for the last 5 years and is duplicated across different systems I have...
So I always push to `fork` while pulling from `origin` upstream, unless `origin` is where I work directly, that is.

Reply:
Maybe, we could implement having a setting in `.git/config`? I've implemented some helpers for this @ https://github.com/python/cherry-picker/blob/2c1f6b9/cherry_picker/cherry_picker.py#L995-L1022.

Reply:
@webknjaz There's no issue with the `origin` remote pointing upstream to the `ansible/ansible` repo. Based on your description, it sounds like you never push to your `devel` branch, since that wouldn't go to your fork. Is that correct?

Reply:
Sometimes I sync the fork via something like `git fetch --all && git push fork origin/devel:devel`.